### PR TITLE
Create features/DictionaryExpression branch and resurrect parsing for those expressions

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This reverts commit 3720edc0a1bae336e58212cdc98b3a01286b9c56.

Creating a new branch for dictionary expressions to live in.  And restoring parsing of KeyValuePairElements.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@cston @RikkiGibson ptal.

**Comment by CyrusNajmabadi** [comment]
@cston @RikkiGibson ptal.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -5199,11 +5199,19 @@ private BoundExpression BindCollectionExpression(CollectionExpressionSyntax synt
 
             MessageID.IDS_FeatureCollectionExpressions.CheckFeatureAvailability(diagnostics, syntax, syntax.OpenBracketToken.GetLocation());
 
-            var builder = ArrayBuilder<BoundNode>.GetInstance(syntax.Elements.Count);
             foreach (var element in syntax.Elements)
             {
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:5203]: Please revert the formatting change.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -5199,11 +5199,19 @@ private BoundExpression BindCollectionExpression(CollectionExpressionSyntax synt
...
-                builder.Add(bindElement(element, diagnostics, this, nestingLevel));
+                if (element is KeyValuePairElementSyntax keyValuePairElement)
+                {
+                    MessageID.IDS_FeatureDictionaryExpressions.CheckFeatureAvailability(diagnostics, syntax, keyValuePairElement.ColonToken.GetLocation());
+                    return new BoundBadExpression(syntax, LookupResultKind.Empty, symbols: [], childBoundNodes: [], CreateErrorType());
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:5207]: >return new BoundBadExpression [](http://example.com/codeflow?start=20&length=29)

We're not reporting any error here currently, although we are generating a BoundTree that cannot be emitted. Consider reporting an existing diagnostic, perhaps:
```csharp
diagnostics.Add(ErrorCode.ERR_ExpressionExpected, syntax.Location);
```

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:5203]: done.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 3]
```diff
@@ -5199,11 +5199,24 @@ private BoundExpression BindCollectionExpression(CollectionExpressionSyntax synt
...
+                }
+            }
+
             var builder = ArrayBuilder<BoundNode>.GetInstance(syntax.Elements.Count);
             foreach (var element in syntax.Elements)
             {
                 builder.Add(bindElement(element, diagnostics, this, nestingLevel));
             }
+
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:5219]:
```suggestion
```

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**